_posts/2015-11-16-CodeRefacorting.md

```diff
@@ -0,0 +1,24 @@
+---
+layout: post
+title: Code Refactoring & Reworking in Waterbear
+author: Adrienne Pind
+---
+
+Refactoring code is an important step in the development process, but can get pretty complicated when working on an open source project. When the contributors on a project are constantly changing, it is common for code to stay in its initial state for a long time. As the code base continues to grow and evolve, this old code can become outdated. This is when refactoring comes into play. Code refactoring can be many things, such as removing duplicate code, improving efficiency of code, reducing bloated functions and adding documentation by making code more readable.
+
+Sometimes refactoring code is not enough to deal with outdated code. Typically refactoring code does not change the functionality of the program, but as any developer can tell you, a project's requirements will always change over time, therefore the code will need to be reworked! In this case "reworking" is just a broader term for refactoring. Refactoring is focused on improving the quality of the code itself, while keeping the functionality the same. Reworking can be anything from completely rewriting code to taking existing code and tweaking it to get a new desired functionality.
+
+Refactoring and reworking code are both essential to keeping a good code base and having a functioning program, but are used at different times. For example, in an open source project such as Waterbear, where contributors are always different, it is very easy for duplicate code to be produced. Sometimes it may be because the code base is large and a new contributor doesn't know where to find a function they need, so they write one instead. Other times it can simply be because they were in a rush and wanted slightly different functionality from an existing implementation so they wrote a new, very similar function. In either case, the duplicate code can be refactored so that there is only one version of the function that encompasses all the previous functions' abilities. But be careful, sometimes this can result in a bloated function! A bloated function is a function that is very long and does many things. Once again this can be solved by refactoring! The bloated function can be separated into multiple functions, each with a specific task, that can then be called to perform the desired functionality. With this sort of refactoring, there is no change in functionality but the code has been changed to be more organized and easier to use and maintain in the future!
```

Comment:
Contributors aren't *always* different. Besides me, there are contributors who have been around for years.